server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/NodeReplaceOverrideWrapper.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation.decider;
+
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+
+/**
+ * A transparent wrapper around an {@link AllocationDecider} that overrides the
+ * provided decider in specific cases to allow node replacement to override
+ * allocation deciders that would otherwise prevent the shard from moving to
+ * the replacement target
+ */
+public class NodeReplaceOverrideWrapper extends AllocationDecider {
+
+    private final AllocationDecider original;
+
+    public NodeReplaceOverrideWrapper(AllocationDecider original) {
+        this.original = original;
+    }
+
+    public AllocationDecider getOriginal() {
+        return original;
+    }
+
+    @Override
+    public Decision canRebalance(ShardRouting shardRouting, RoutingAllocation allocation) {
+        return original.canRebalance(shardRouting, allocation);
+    }
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        final Metadata metadata = allocation.metadata();
+        // If the node is on the source node and moving to the target node,
+        // always allow it, regardless of what the wrapped decider would say.
+        // This overrides things like disk space, etc.
+        if (isReplacementTarget(metadata, node.nodeId()) && isReplacementSource(metadata, shardRouting.currentNodeId())) {
```

Comment:
I think we need a more one-to-one mapping between replacing nodes, since otherwise we risk that replacing two nodes concurrently moves all the large shards to one target and all the small ones to another target if we are unlucky.

Or just shuffles the tiers around, allows ccr'ing on non remote nodes etc.

Reply:
Yeah good catch :+1: